README.md
RecSync
=======

The record synchronizer project includes two parts.
A client (RecCaster) which runing as part of an EPICS
IOC, and a server (RecCeiver) which is a stand alone
daemon.  Together they work to ensure the the server(s)
have a complete list of all records currently provided
by the client IOCs.

Information Uploaded
--------------------

The RecCaster client currently sends the following information
about its Process Database.

* The EPICS Base Version
* A white listed set of envronment variables
* The name and type of all records
* Any info() tags associated with these records

RecCaster Usage
---------------

The RecCaster source in the `client/` sub-directory
is build as an EPICS support module.
It provides `reccaster.dbd` and the `reccaster` library.
The client demo IOC in `client/demoApp/` provides
an example of how to build RecCaster into an IOC.

RecCaster requires only EPICS Base 3.15 and later 3.14 releases.

RecCeiver Usage
---------------

The RecCeiver server in the `server/` directory is a
Python script using the [Twisted][twisted] networking
library.  It requires Python 2.7 and Twisted >= 12.0.

[twisted]: http://twistedmatrix.com/

The server uses the Twisted plugin interface to
make client information available to one or more
plugins.  See `server/demo.conf` for an example
configuration.

Currently two plugins are provided: `show` which print client
information to screen/log, and `db` which writes into a SQL
database (currently only sqlite3 supported).
The SQL table schema used is defined in `server/recceiver.sqlite3`.

Theory of Operation
===================

RecCaster
---------

The RecCaster client is an epics support module
which automatically starts a single thread and
begins passively waiting for a RecCeiver server
to announce its presence.

Upon receiving an announcement the client will
connect and upload its list of records and related
information to the server.  It then remains
connected indefinately and responds to periodic
alive tests from the server.

If this connection is lost the client will resume
waiting for another announcement.

RecCeiver
---------

The RecCeiver server periodially (every 15 seconds) sends
a broadcast announcement of it existance.

When a client connects it will wait for a server
handshake message before beginning to upload records.
The server will delay sending this message if too
many clients are currently uploading.

UDP Protocol
------------

Server announcement are made to UDP port 5049.

When sent to a IPv4 address,
such messages will have a length >= 16 bytes.
Clients will process only the first 16 bytes
and ignore any remaining bytes.

<table border="1">
  <tr>
    <td>0</td>
    <td>1</td>
    <td>2</td>
    <td>3</td>
    <td>4</td>
    <td>5</td>
    <td>6</td>
    <td>7</td>
    <td>8</td>
    <td>9</td>
    <td>A</td>
    <td>B</td>
    <td>C</td>
    <td>D</td>
    <td>E</td>
    <td>F</td>
  </tr>
  <tr>
    <td colspan="2">ID</td>
    <td>0</td>
    <td>X</td>
    <td colspan="4">SERV ADDR</td>
    <td colspan="2">PORT</td>
    <td>X</td>
    <td>X</td>
    <td colspan="4">SERV KEY</td>
  </tr>
</table>

* All multibyte fields have big-endian ordering.
* Fields marked `0` must be zero and clients will (silently) reject
messages where these fields are not zero.
* Fields marked `X` should be ignored by clients.
* The `ID` field must have the value 0x5243 (ascii 'RC').
* The `SERV ADDR` and `PORT` fields is the IPv4 address and TCP port to which clients should connect.
* `SERV KEY` is an arbitrary number which the client will later on the TCP connection.


TCP Protocol
------------

The TCP data stream will consist of a sequence of messages
having an 8 byte header followed by a variable length body.
The header will have the form

<table border="1">
  <tr>
    <td>0</td>
    <td>1</td>
    <td>2</td>
    <td>3</td>
    <td>4</td>
    <td>5</td>
    <td>6</td>
    <td>7</td>
  </tr>
  <tr>
    <td colspan="2">ID</td>
    <td colspan="2">MSGID</td>
    <td colspan="4">LEN</td>
  </tr>
</table>

* All multibyte fields have big-endian ordering.
* The `ID` field must have the value 0x5243 (ascii 'RC').
* The `MSGID` field identifies the type of message to follow.
* The `LEN` field gives the length in bytes or by message body to follow the header.
* Messages from server to client will have `MSGID` >= 0x8000.
* Messages from client to server will have `MSGID` < 0x8000.
* Except where explicitly stated, endpoints must silently ignore messages with unknown `MSGID`
* Endpoints must be prepared to receive and ignore extra body bytes.

*Messages*

<table border="1">
 <tr>
  <th>MSGID</th>
  <th>Min. LEN</th>
  <th>Name</th>
  <th>0</th>
  <th>1</th>
  <th>2</th>
  <th>3</th>
  <th>4</th>
  <th>5</th>
  <th>6</th>
  <th>7</th>
 </tr>
 <tr>
  <td>0x8001</td>
  <td>1</td>
  <td>Server Greet</td>
  <td>0</td>
 </tr>
 <tr>
  <td>0x8002</td>
  <td>4</td>
  <td>Ping</td>
  <td colspan="4">NONCE</td>
 </tr>
 <tr>
  <td>0x0001</td>
  <td>8</td>
  <td>Client Greet</td>
  <td>0</td>
  <td>0</td>
  <td>X</td>
  <td>X</td>
  <td colspan="4">SERV KEY</td>
 </tr>
 <tr>
  <td>0x0002</td>
  <td>4</td>
  <td>Pong</td>
  <td colspan="4">NONCE</td>
 </tr>
 <tr>
  <td>0x0003</td>
  <td>9</td>
  <td>Add Record</td>
  <td colspan="4">RECID</td>
  <td>ATYPE</td>
  <td>RTLEN</td>
  <td colspan="2">RNLEN</td>
  <td>RTYPE</td>
  <td>RNAME</td>
 </tr>
 <tr>
  <td>0x0004</td>
  <td>4</td>
  <td>Del Record</td>
  <td colspan="4">RECID</td>
 </tr>
 <tr>
  <td>0x0005</td>
  <td>4</td>
  <td>Upload Done</td>
  <td>0</td>
  <td>0</td>
  <td>0</td>
  <td>0</td>
 </tr>
 <tr>
  <td>0x0006</td>
  <td>9</td>
  <td>Add Info</td>
  <td colspan="4">RECID</td>
  <td>KEYLEN</td>
  <td>X</td>
  <td colspan="2">VALEN</td>
  <td>KEY</td>
  <td>VALUE</td>
 </tr>
</table>

* `NONCE` is an number choosen by the server for a Ping message.  It must be echoed back by the client in a Pong message.
* `SERV KEY` must be the number the client received in the announcement from which it found this server.
* `RECID` is an identifier choosen by the client to identify a single record instance.  It must be >0.
* `ATYPE` is 0 to add a record entry or 1 to add a record alias.  When adding an alias the record type is omitted (`RTLEN`==0) and the `RECID` must have been added in a previous message with `ATYPE`==0.
* `RTLEN`, `RNLEN`, `KEYLEN`, and `VALEN` are lengths in bytes.
* `RTLEN` and `VALEN` are allowed to be zero.
* `RNLEN` and `KEYLEN` must be greater than zero.
* `RTYPE`, `RNAME`, `KEY`, and `VALUE` are variable length fields whos length is given by the corresponding `*LEN` field.
* `RTYPE`, `RNAME`, `KEY`, and `VALUE` should not include a 0 (ascii null) byte.
* The `RECID` field of the Add Info may be zero.  When zero the key/value pair is considered to be associated with the client as a whole, and not any individual record.

When a TCP connection is established the client must send the Client Greeting message
immediately, then wait to receive the Server Greeting message before sending any
further messages.

Once the client has recevieved the Server Greeting it may send begin sending Add Record, Add Info, and Del Record messages.
Once the initial upload has been completed, the client must send Upload Done.

After Upload Done is received the server will periodically send Ping messages.  The client must respond to each Ping with a Pong.
If a client does not respond promptly the server may close the connection.


debian/changelog
recsync (1.3.6) unstable; urgency=low

  * Send PING message immediately when DONE message is received
  * Commit transactions sequentially and support cancellation
  * Collect and log more performance metrics
  * Remove the extra newline character from the log messages
  * Add support for cfstore commit cancellation
  * Allow the ChannelFinder client size limit to be configured
  * Improve cfstore performance by hashing PV info by PV name instead of RID

 -- Dylan Maxwell <maxwelld@frib.msu.edu>  Fri, 14 Mar 2019 10:11:12 -0400

recsync (1.3.5) unstable; urgency=medium

  * Fix bug with client that caused RIDs to be reused when more than 65536 records

 -- Dylan Maxwell <maxwelld@frib.msu.edu>  Thu, 27 Feb 2019 17:25:14 -0400

recsync (1.3.4) unstable; urgency=low

  * Record Alias in CF

 -- Tanvi Ashwarya <ashwarya@frib.msu.edu>  Thu, 27 Dec 2018 09:46:18 -0400

recsync (1.3.3) unstable; urgency=low

  * BUG: set arithmetic

 -- Daron Chabot <chabot@frib.msu.edu>  Mon, 30 Oct 2017 15:54:09 -0400

recsync (1.3.2) unstable; urgency=low

  * BLD: use setuptools instead of distutils
  * WIP: whitelist for CF info tags
  * WIP: placeholder
  * ENH: add whitelisted infotag to CF properties
  * DOC: comment, example of infotag whitelist conf
  * ENH: sanitize whitelist
  * CLN: remove general properties addition mechanism
  * WIP: property cacheing at init
  * CLN: hitting service per ioc
  * FIX: hang up if no one is home
  * MNT: clean up
  * MNT: CF dependency is optional

 -- Daron Chabot <chabot@frib.msu.edu>  Fri, 27 Oct 2017 17:19:03 -0400

recsync (1.3.1) unstable; urgency=medium

  * Fix uninitialized variable in client

 -- Dylan Maxwell <maxwelld@frib.msu.edu>  Wed, 09 Nov 2016 15:52:41 -0500

recsync (1.3) unstable; urgency=medium

  * Update cf-store with improved error handling and info tag support

 -- Dylan Maxwell <maxwelld@frib.msu.edu>  Mon, 07 Nov 2016 16:36:58 -0500

recsync (1.2) unstable; urgency=medium

  * Add in cf-store plugin
  * reccaster sends HOSTNAME
  * Fix receiver PID file deletion
  * add 'loglevel' config option
  * set SO version '1' in debian/rules

 -- Michael Davidsaver <mdavidsaver@gmail.com>  Wed, 27 Jul 2016 13:20:35 -0400

recsync (1.1) unstable; urgency=medium

  [ dylan171 ]
  * Error using demo configuration file
  * System configuration clobbers local configuration

  [ Michael Davidsaver ]
  * ConfigParser read([]) allows later entries to override previous
  * add top level license and copyright info

  [ Daron Chabot ]
  * FIX: support full installation
  * ENH: permit override of logfile location

  [ Michael Davidsaver ]
  * no install failure if daemon doesn't start

 -- Michael Davidsaver <mdavidsaver@gmail.com>  Wed, 01 Jun 2016 17:58:53 -0400

recsync (1.0) unstable; urgency=medium

  * initial

 -- Michael Davidsaver <mdavidsaver@gmail.com>  Mon, 11 Apr 2016 18:04:35 -0400


client/castApp/src/caster.h
#ifndef CASTER_H
#define CASTER_H

#include <epicsTypes.h>
#include <osiSock.h>
#include <epicsEvent.h>
#include <epicsMutex.h>
#include <compilerDependencies.h>

#include "sockhelpers.h"

#define RECAST_PORT 5049

epicsShareExtern double reccastTimeout;
epicsShareExtern double reccastMaxHoldoff;

typedef enum {
  casterUDPSetup,
  casterTCPSetup,
} caster_h;

typedef enum {
    casterStateInit,
    casterStateListen,
    casterStateConnect,
    casterStateUpload,
    casterStateDone,
} casterState;

typedef struct _caster_t {
    double timeout;

    unsigned short udpport;

    ssize_t nextRecID;
    shSocket *csock;

    /* doCaster*Phase() listens on [1] */
    SOCKET wakeup[2];

    int errors;
    epicsEventId shutdownEvent;

    int haveserv;
    osiSockAddr nameserv;
    epicsUInt32 servkey;

    /* overwritable hook functions */
    void (*testhook)(struct _caster_t*, caster_h);

    void *arg;
    int (*getrecords)(void*, struct _caster_t*);
    void (*onmsg)(void*, struct _caster_t*);

    casterState current;

    /* lock for message and shutdown flags */
    epicsMutexId lock;

    int shutdown;
    char lastmsg[MAX_STRING_SIZE];
} caster_t;

epicsShareFunc
void casterMsg(caster_t *self, const char* msg, ...) EPICS_PRINTF_STYLE(2,3);

#define ERRTO(TAG, CASTER, ...) do{ casterMsg(CASTER, __VA_ARGS__); goto TAG; }while(0)

#define ERRRET(VAL, CASTER, ...) do{ casterMsg(CASTER, __VA_ARGS__); return (VAL); }while(0)

epicsShareFunc
void casterInit(caster_t *self);
epicsShareFunc
void casterShutdown(caster_t *self);
epicsShareFunc
int casterStart(caster_t *self);

epicsShareFunc
ssize_t casterSendRecord(caster_t* c, const char* rtype, const char* rname);
epicsShareFunc
ssize_t casterSendAlias(caster_t* c, size_t rid, const char* rname);
epicsShareFunc
int casterSendInfo(caster_t *c, ssize_t rid, const char* name, const char* val);

/* push process database information */
epicsShareFunc
int casterPushPDB(void *junk, caster_t *caster);

/* internal */

epicsShareFunc
int doCasterUDPPhase(caster_t *self);
epicsShareFunc
int doCasterTCPPhase(caster_t *self);

/* recv() a TCP protocol message.
 * Message id is stored in *id.
 * returns the number of body bytes stored in the buffer.
 * If a message w/ zero body length is recieved this
 * is reported as a disconnect.
 *
 * Unless MSG_PARTIAL is given, message bodies too long
 * to find in the provided buffer will trigger an error.
 */
epicsShareFunc
ssize_t casterRecvPMsg(shSocket* s, epicsUInt16* id,
                       void *buf, size_t len, int flags);

/* recv() a TCP protocol header.
 *
 * returns 1 on success, 0 when connection is lost, and -1 on error
 */
epicsShareFunc
int casterRecvPHead(shSocket *s, epicsUInt16* id, epicsUInt32* len, int flags);

epicsShareFunc
int casterSendPHead(shSocket* s, epicsUInt16 id, epicsInt32 blen);

#define RECAST_MAGIC 0x5243 /* 'RC' */

/* UDP server advertisement message
 *
 *    | 0 | 1 | 2 | 3 |
 * 00 | ID    | 0 | X |
 * 04 |   SERV ADDR   |
 * 08 | PORT  | X | X |
 * 0C |   SERV KEY    |
 * 10 |   ignore...
 */
typedef struct {
    epicsUInt16 pid; /* protocol ID RECAST_MAGIC */
    epicsUInt8 version; /* must be 0 */
    epicsUInt8 reserved0;
    epicsUInt32 serverIP;
    epicsUInt16 serverPort;
    epicsUInt16 reserved1;
    epicsUInt32 serverKey;
} casterAnnounce;

union casterUDP {
    casterAnnounce m_msg;
    char m_bytes[sizeof(casterAnnounce)];
};

/* TCP message header
 *
 *    | 0 | 1 | 2 | 3 |
 * 00 |  PID  | MSGID |
 * 04 |  body length  |
 * 08 | body bytes...
 */
typedef struct {
    epicsUInt16 pid; /* protocol ID RECAST_MAGIC */
    epicsUInt16 msgid;
    epicsUInt32 bodylen;
} casterHeader;

union casterTCPHead {
    casterHeader m_msg;
    char m_bytes[sizeof(casterHeader)];
};

/* server messages */
typedef struct {
    epicsUInt8 version;
} casterServerGreet; /* 0x8001 */

typedef struct {
    epicsUInt32 nonce;
} casterPing; /* 0x8002 and 0x0002 */

/* client messages */
typedef struct {
    epicsUInt8 version;
    epicsUInt8 type; /* 0 - static, 1 - dynamic, >127  - Site specific */
    epicsUInt8 reserved[2];
    epicsUInt32 serverKey;
} casterClientGreet; /* 0x0001 */

typedef struct {
    epicsUInt32 rid; /* record instance id */
    epicsUInt8 rtype; /* 0 - IOC Rec, 1 - IOC Alias */
    epicsUInt8 rtlen; /* # of bytes in record type name (0 for aliases) */
    epicsUInt16 rnlen; /* # of bytes in record instance name */
    /* record type and instance names follow */
} casterClientAddRec; /* 0x0003 */

typedef struct {
    epicsUInt32 rid; /* record instance id */
} casterClientDelRec; /* 0x0004 */

/* casterClientDone 0x0005 dummy 4 byte payload */

typedef struct {
    epicsUInt32 rid; /* record instance id */
    epicsUInt8 klen; /* # of bytes in record type name */
    epicsUInt8 reserved;
    epicsUInt16 vlen; /* # of bytes in record instance name */
    /* key and value strings follow */
} casterClientAddInfo; /* 0x0006 */

union casterTCPBody {
    casterServerGreet s_greet;
    casterPing ping;

    casterClientGreet c_greet;
    casterClientAddRec c_add;
    casterClientDelRec c_del;
    casterClientAddInfo c_info;

    char bytes[sizeof(casterClientAddRec)];
};


#endif // CASTER_H
